lib/HLSL/HLOperationLower.cpp

```diff
@@ -6067,6 +6068,74 @@ Value *GEPIdxToOffset(GetElementPtrInst *GEP, IRBuilder<> &Builder,
+// If we are provided a non-zero offset, we need to simulate it
+// by returning the correct elements.
+using ResRetValueArray = std::array<Value*, 4>;
+static ResRetValueArray GenerateTypedBufferLoad(
+  Value *Handle, Type *BufferElemTy, Value *ElemIdx, Value *StatusPtr,
+  OP* HlslOP, IRBuilder<> &Builder) {
+
+  OP::OpCode OpCode = OP::OpCode::BufferLoad;
+  Value* LoadArgs[] = { HlslOP->GetU32Const((unsigned)OpCode), Handle, ElemIdx, UndefValue::get(Builder.getInt32Ty()) };
+  Function* LoadFunc = HlslOP->GetOpFunc(OpCode, BufferElemTy);
+  Value* Load = Builder.CreateCall(LoadFunc, LoadArgs, OP::GetOpCodeName(OpCode));
+
+  ResRetValueArray ResultValues;
+  for (unsigned i = 0; i < ResultValues.size(); ++i) {
+    ResultValues[i] = cast<ExtractValueInst>(Builder.CreateExtractValue(Load, { i }));
+  }
+
+  UpdateStatus(Load, StatusPtr, Builder, HlslOP);
+  return ResultValues;
+}
+
+static AllocaInst* SpillValuesToArrayAlloca(ArrayRef<Value*> Values, IRBuilder<>& Builder) {
+  DXASSERT_NOMSG(!Values.empty());
+  IRBuilder<> AllocaBuilder(dxilutil::FindAllocaInsertionPt(Builder.GetInsertPoint()));
+  AllocaInst* ArrayAlloca = AllocaBuilder.CreateAlloca(ArrayType::get(Values[0]->getType(), Values.size()));
+  for (unsigned i = 0; i < Values.size(); ++i) {
+    Value* ArrayElemPtr = Builder.CreateGEP(ArrayAlloca, { Builder.getInt32(0), Builder.getInt32(i) });
+    Builder.CreateStore(Values[i], ArrayElemPtr);
+  }
+  return ArrayAlloca;
+}
+
+static Value* ExtractFromTypedBufferLoad(const ResRetValueArray& ResRet,
+    Type* ResultTy, Value* Offset, IRBuilder<>& Builder) {
+  unsigned ElemCount = ResultTy->isVectorTy() ? ResultTy->getVectorNumElements() : 1;
+  DXASSERT_NOMSG(ElemCount < ResRet.size());
+  unsigned ElemSizeInBytes = ResRet[0]->getType()->getScalarSizeInBits() / 8;
+
+  SmallVector<Value*, 4> Elems;
+  if (ConstantInt *OffsetAsConstantInt = dyn_cast<ConstantInt>(Offset)) {
+    // Get all elements to be returned
+    uint64_t FirstElemOffset = OffsetAsConstantInt->getLimitedValue();
+    DXASSERT_NOMSG(FirstElemOffset % ElemSizeInBytes == 0);
+    uint64_t FirstElemIdx = FirstElemOffset / ElemSizeInBytes;
+    DXASSERT_NOMSG(FirstElemIdx < ResRet.size() - ElemCount);
+    for (unsigned ElemIdx = 0; ElemIdx < ElemCount; ++ElemIdx) {
+      Elems.emplace_back(ResRet[std::min<size_t>(FirstElemIdx + ElemIdx, ResRet.size() - 1)]);
+    }
+  }
+  else {
+    Value* ArrayAlloca = SpillValuesToArrayAlloca(
+      ArrayRef<Value*>(ResRet.data(), ResRet.size()), Builder);
+
+    // Get all elements to be returned through dynamic indices
+    Value *FirstElemIdx = Builder.CreateUDiv(Offset, Builder.getInt32(ElemSizeInBytes));
+    for (unsigned i = 0; i < ElemCount; ++i) {
+      Value *ElemIdx = Builder.CreateAdd(FirstElemIdx, Builder.getInt32(i));
+      Value* ElemPtr = Builder.CreateGEP(ArrayAlloca, { Builder.getInt32(0), ElemIdx });
+      Elems.emplace_back(Builder.CreateLoad(ElemPtr));
+    }
```

Comment:
do we have test for dynamic indexing scenario?

Reply:
Now we do!